azure/services/bastionhosts/spec.go

```diff
@@ -40,69 +41,64 @@ type AzureBastionSpec struct {
 	EnableTunneling bool
 }
 
-// AzureBastionSpecInput defines the required inputs to construct an azure bastion spec.
-type AzureBastionSpecInput struct {
-	SubnetName   string
-	PublicIPName string
-	VNetName     string
-}
-
-// ResourceName returns the name of the bastion host.
-func (s *AzureBastionSpec) ResourceName() string {
-	return s.Name
-}
-
-// ResourceGroupName returns the name of the resource group.
-func (s *AzureBastionSpec) ResourceGroupName() string {
-	return s.ResourceGroup
-}
-
-// OwnerResourceName is a no-op for bastion hosts.
-func (s *AzureBastionSpec) OwnerResourceName() string {
-	return ""
+// ResourceRef implements azure.ASOResourceSpecGetter.
+func (s *AzureBastionSpec) ResourceRef() *asonetworkv1.BastionHost {
+	return &asonetworkv1.BastionHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.Name,
+			Namespace: s.Namespace,
+		},
+	}
 }
 
-// Parameters returns the parameters for the bastion host.
-func (s *AzureBastionSpec) Parameters(ctx context.Context, existing interface{}) (parameters interface{}, err error) {
-	if existing != nil {
-		if _, ok := existing.(armnetwork.BastionHost); !ok {
-			return nil, errors.Errorf("%T is not an armnetwork.BastionHost", existing)
-		}
-		// bastion host already exists
-		return nil, nil
+// Parameters implements azure.ASOResourceSpecGetter.
+func (s *AzureBastionSpec) Parameters(ctx context.Context, existingBastionHost *asonetworkv1.BastionHost) (parameters *asonetworkv1.BastionHost, err error) {
+	if existingBastionHost != nil {
+		return existingBastionHost, nil
```

Comment:
Not sure why we want to return `existingBastionHost` early. 
Since we want CAPZ's opinions to supersede user inputs (only on the fields CAPZ has an opinion about), we should instead have something like below at line 62.
```suggestion
	if existingBastionHost != nil {
		bastionHost.Spec = existingBastionHost.Spec
	}
```

Reply:
Won't this actually cause unecessary fields to be populated like an IpConfiguration with empty ARMIDs? https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4143/files#diff-62fea3ceadec31e79fa0bae79a9972f0c16cb5cc274a934828520986acda0a46R81

Reply:
I don't fully understand your statement; _"unecessary fields to be populated like an IpConfiguration with empty ARMIDs"_.

Since 
- CAPZ is populating the spec of the ASO resource (and)
- CAPZ's opinion should be respected over user interventions(where user updates a CAPZ managed ASO resource), 

if the user initially created an IP configuration with Empty ARMIDs, then empty ARMIDs should be populated.

Reply:
Returning the object early if found (as it was implemented earlier) will let the users modify the CAPZ-managed resource, and CAPZ will not be able to enforce its values.

Reply:
That makes sense, I think my confusion is because the previous implementation would get the result as nil if an existing [BastionHost is present with empty data](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/bastionhosts/spec_test.go#L80). But now, we would expect a BastionHost with default empty values like [so](https://github.com/willie-yao/cluster-api-provider-azure/blob/aso-bastionhosts/azure/services/bastionhosts/spec_test.go#L102). It just makes the unit test a bit awkward.